[utils] Add opt-in `DataAttributesOverrides` augmentation for slot props (#48554)
mui-org/material-ui: docs/data/material/guides/typescript/typescript.md

You have to decide for yourself if the same tradeoff is acceptable for you.
The library types are strict by default and loose via opt-in.

## Allowing `data-*` attributes on `slotProps`

By default, slot prop types reject arbitrary `data-*` attributes, even though they are forwarded to the DOM at runtime.
This keeps the typed surface tight and catches typos.

> To pass `data-*` attributes through `slotProps` (for example, test locators like `data-testid`), opt in by augmenting the `DataAttributesOverrides` interface from `@mui/utils/types` once, anywhere in your project:

Reviewer: ```suggestion
Augment the `DataAttributesOverrides` interface to specify `data-*` attributes on `slotProps`:
```

to simplify.

Author: Thanks. Simplified. 👌

```ts
// Accept any data-* attribute on every slot.

> declare module '@mui/utils/types' {

Reviewer: ```suggestion
declare module '@mui/material/utils' {
```

Can it be reexport so that user don't need to care about `@mui/utils`?

Author: Re-exported and updated the guide example: https://deploy-preview-48554--material-ui.netlify.app/material-ui/guides/typescript/#allowing-data-attributes-on-slotprops.
